server/resp/src/main/java/org/infinispan/server/resp/RespLettuceHandler.java

```diff
@@ -19,18 +22,53 @@ public class RespLettuceHandler extends ByteToMessageDecoder {
 
    private final RedisStateMachine stateMachine = new RedisStateMachine(ByteBufAllocator.DEFAULT);
    private RespRequestHandler requestHandler;
+   private boolean disabledRead = false;
 
    public RespLettuceHandler(RespRequestHandler initialHandler) {
       this.requestHandler = initialHandler;
    }
 
    @Override
    protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) {
+      // Don't read any of the ByteBuf if we disabled reads
+      if (disabledRead) {
+         return;
+      }
       PushOutput<byte[], byte[]> pushOutput = new PushOutput<>(ByteArrayCodec.INSTANCE);
       if (stateMachine.decode(in, pushOutput)) {
          String type = pushOutput.getType().toUpperCase();
          List content = pushOutput.getContent();
-         requestHandler = requestHandler.handleRequest(ctx, type, content.subList(1, content.size()));
+         List<byte[]> contentToUse = content.subList(1, content.size());
+         log.tracef("Received command: %s with arguments %s", type, Util.toStr(contentToUse));
+         CompletionStage<RespRequestHandler> stage = requestHandler.handleRequest(ctx, type, contentToUse);
+         if (CompletionStages.isCompletedSuccessfully(stage)) {
+            requestHandler = CompletionStages.join(stage);
+         } else {
+            log.tracef("Disabling auto read for channel %s until previous command is complete", ctx.channel());
```

Comment:
This is where the backpressure for netty comes in. Basically if the command hasn't completed yet, we disable netty from reading more bytes from the socket `setAutoRead(false)` and set a variable to prevent additional commands from being processed `disabledRead = true;`. Then after the command completes we allow netty to read on the socket and for us to process commands.

We have to also finally check if there are any additional bytes read from the socket already in the netty buffer and invoke decode if so. This way if read 2 commands from the socket and the first doesn't complete right away we can process the second command without any additional activity on the socket.